### ISSUE TITLE
Defining custom coercions from extant types

### ISSUE THREAD

**Issue by alabamenhu** [Raku/problem-solving#22]
I remember this being discussed a while back, and I may throw a prototype together at some point but figured I'd propose it here first.

Right now, classes can define how they coerce into other classes which is particularly useful in signatures, but also in other general cases.

```
class A {
  method Str { ... } # coerces to string 
  method Int { ... } # coerces to int
}
```

This coercion for many built ins is two way:

```
"123".Int.Str # from string to integer and back again
```

Suppose our class A can be created from a Str:

```
class A { 
   multi method new (Str $foo) { … }
   method Str { … }
}
```

It is possible to round trip only one way:

```
my $a = A.new;
A.new($a.Str); # good
$a.Str.A; # error, Str has no method A
```

The chaining method structure (`A.new.Str.A`) is not available because that would require augmenting Str (something that's possible, but highly discouraged).  While fairly easy to work around in general code, it does prevent us from passing a `Str` to a sub whose signature allows for `A()`.  Even where we can define each class in full, because of single-pass parsing, it makes doing something like the following quite difficult

```
class A {
  method B { … }
}
class B {
  method A { ... } 
}
```

I would propose having a standardized method akin to `ACCEPTS` by which classes can define how *others* classes coerce into them.  Thus the above cyclic problem could be at least partially solved by only needing one of them to defined the other:

```
class A { ... }
class B { 
  method A { A.new: … }
  method FROM(A:D $a) { self.bless: … }
}
```

When type checking the signature in signature like `sub foo (B() $b) { … }` the process would then flow as 

  1. Is `$b` an `B`? If so, successful signature check.
  2. If not, does `$b` have the method `.B`?  If so, `$b = $b.B` and successful signature check.
  3. If not, does `B` have the method `FROM(Type:D)` where Type is any of the types in `$b.^mro`?  If so, `$b = B.FROM($b)` and successful signature check.
  4. Fail signature check.

While this would work work fantastically for signatures, I'm not sure it would be something that should be automatically tried in method chaining (e.g. `A.new.B` might want to fail, but a quick sub in the similar to the following would handle it whenever there were any question.

```
sub coerce ($from, $to) {
  $from.?"{$to.^name}"() 
    // $to.?FROM($from) 
    // die "Cannot coerce $from into {$to.WHAT}";
}
```

(Obviously, it'd be a bit more complex than that, as if a coercion fails, you'd want to simply try the next one along the ^mro chain).

The utility of such a method would be pretty tangible particularly whenever a custom class can easily be obtained from anything of type `Cool`.  In quite a few of my `Intl` modules, a `LanguageTag` object is one of the argument types.  A lot of times it's more convenient for people to use pass in a `Str` directly that then is coerced into a `LanguageTag` object.  The result is a lot of `multi`s to handle the coercion which the idea of `Type() $foo` signature structure is supposed to help avoid.

**Comment by alabamenhu** [comment]
Apparently according to this [gist](https://gist.github.com/skids/aabd2aad3d0b5ad8481b), the spec originally called for the structure to be 

```
class A { ... }
class B {
  method A { … } # coerce into A
  multi method new (A:D $a) { self.bless: ... } #coerce from A
}
```

But that this wasn't hasn't implemented for some number of reasons.

**Comment by vrurg** [comment]
Closed via rakudo/rakudo#3891
